### ISSUE TITLE
server memory leak when client directly joins with network-ai=server-max-players

### ISSUE THREAD

**Issue by hugeblank** [supertuxkart/stk-code#5138]
## Description
When a client on localhost (or anywhere if `ai-anywhere` is true!) attempts to join a server with `--network-ai` equal to the max amount of karts, the server chews through memory like there's no tomorrow. The leak persists after the player disconnects.

Setting `--network-ai` to any amount less than the max does not cause this issue.

## Steps to reproduce
1. Set up server as usual following [NETWORKING.md](https://github.com/supertuxkart/stk-code/blob/master/NETWORKING.md). My specific instance has a password and disables polling on the command line, though I doubt this is relevant.
2. Join a client with: `supertuxkart --connect-now=localhost --server-password=<password> --network-ai=8`
3. Keep an eye on system memory on the server. In my tinkering it went up at an alarming 600MB/s. 

## Configuration

STK release version: 
- server: e6d0df3
- client: 1.4

STK Source: 
- server: git
- client: arch packages [`extra`](https://archlinux.org/packages/extra/x86_64/supertuxkart/)

System: Arch Linux 6.10.2-arch1-1

Graphics card: Radeon RTX 6600 XT

CPU: Ryzen 5800 X3D

## Additional information
stdout.log (does not appear to be relevant):
```
[error  ] User Config: Unknown value for non_ge_fullscreen_desktop; expected true or false
[error  ] User Config: Couldn't find parameter group server-bookmarks-order
[info   ] [FileManager]: Data files will be fetched from: '../data/'
[info   ] [FileManager]: Data files will be fetched from: '../data/../../stk-assets/'
[info   ] [FileManager]: User directory is '/home/hugeblank/.config/supertuxkart/config-0.10/'.
[info   ] [FileManager]: Addons files will be stored in '/home/hugeblank/.local/share/supertuxkart/addons/'.
[info   ] [FileManager]: Screenshots will be stored in '/home/hugeblank/.local/share/supertuxkart/screenshots/'.
[info   ] [FileManager]: User-defined grand prix will be stored in '/home/hugeblank/.local/share/supertuxkart/grandprix/'.
[info   ] [FileManager]: Asset 0 will be loaded from '../data/challenges/'.
[info   ] [FileManager]: Asset 1 will be loaded from '../data/gfx/'.
[info   ] [FileManager]: Asset 2 will be loaded from '../data/grandprix/'.
[info   ] [FileManager]: Asset 3 will be loaded from '../data/gui/icons/'.
[info   ] [FileManager]: Asset 4 will be loaded from '../data/gui/screens/'.
[info   ] [FileManager]: Asset 5 will be loaded from '../data/gui/dialogs/'.
[info   ] [FileManager]: Asset 6 will be loaded from '../data/replay/'.
[info   ] [FileManager]: Asset 7 will be loaded from '../data/shaders/'.
[info   ] [FileManager]: Asset 8 will be loaded from '../data/skins/'.
[info   ] [FileManager]: Asset 9 will be loaded from '../data/ttf/'.
[info   ] [FileManager]: Asset 10 will be loaded from '../data/po/'.
[info   ] [FileManager]: Asset 11 will be loaded from '../data/../../stk-assets/library/'.
[info   ] [FileManager]: Asset 12 will be loaded from '../data/../../stk-assets/models/'.
[info   ] [FileManager]: Asset 13 will be loaded from '../data/../../stk-assets/music/'.
[info   ] [FileManager]: Asset 14 will be loaded from '../data/../../stk-assets/sfx/'.
[info   ] [FileManager]: Asset 15 will be loaded from '../data/../../stk-assets/textures/'.
[info   ] [FileManager]: Asset 16 will be loaded from '../data/../../stk-assets/tracks/'.
[info   ] main: STK config will be read from stk_config.xml.
[error  ] skin: Could not read XML file '../data/skins/peach/stkskin.xml'.
[error  ] GUI: Could not load skin, reverting to default.
[info   ] GUI: Loading skin data from file: ../data/skins/classic/stkskin.xml
[info   ] HTTPRequest: Sending userid=481947&token=************************ to https://online.supertuxkart.net/api/v2/user/saved-session/
[info   ] GrandPrixManager: Loading Grand Prix files from ../data/grandprix/
[info   ] GrandPrixManager: Loading Grand Prix files from /home/hugeblank/.local/share/supertuxkart/grandprix/
Tue Aug 06 21:11:42 2024 [info   ] STKHost: Host initialized.
Tue Aug 06 21:11:42 2024 [info   ] STKHost: Server port is 2759
Tue Aug 06 21:11:42 2024 [info   ] ServerLobby: Resetting the server to its initial state.
Tue Aug 06 21:11:42 2024 [info   ] ProtocolManager: A 11ServerLobby protocol has been started.
Tue Aug 06 21:11:42 2024 [info   ] main: Creating a LAN server 'STK Server'.
Tue Aug 06 21:11:42 2024 [info   ] Achievements: Synching achievement 2,4,5,7,9 to server.
Tue Aug 06 21:11:42 2024 [info   ] HTTPRequest: Sending userid=481947&token=************************&achievementid=2%2C4%2C5%2C7%2C9 to https://online.supertuxkart.net/api/v2/user/achieving/
Tue Aug 06 21:11:42 2024 [info   ] STKHost: Listening has been started.
Tue Aug 06 21:11:43 2024 [info   ] HTTPRequest: Sending userid=481947&token=************************&visitingid=481947 to https://online.supertuxkart.net/api/v2/user/get-friends-list/
Tue Aug 06 21:11:51 2024 [info   ] STKHost: [::1]:37374 has just connected. There are now 1 peers.
Tue Aug 06 21:11:51 2024 [info   ] ServerLobby: Message of type 1 received.
```

**Comment by Alayan-stk-2** [comment]
If you have two clients each joining with half of the max amount of karts (using an even max number so that together it's the max amount), what happens?

If the issue is related to completely filling a server with AI karts, this should also trigger the leak.

**Comment by hugeblank** [comment]
> If you have two clients each joining with half of the max amount of karts (using an even max number so that together it's the max amount), what happens?

On testing it seems I can't connect multiple `network-ai` clients on the server with the same login (even with less than the max). I tried creating a new account and logging in through that but the currently logged in user in `players.xml` takes precedence over the `--login` command line option, confirmed by cross referencing the player-id that pops up in the log with the already open and connected process (and a lookup using the online dashboard).

I thought the power of friendship might be able to fix this, and so I flipped the `ai-anywhere` value to try and have a friend connect from their account, but the server rejected them. Their log:
```
[info   ] ClientLobby: Synchronous message of type 2
[error  ] STKHost: Connection refused: Invalid player connecting.
[info   ] ProtocolManager: A 11ClientLobby protocol has been terminated.
```
Server log:
```
Sun Aug 11 13:55:36 2024 [info   ] STKHost: [ip]:[port] has just connected. There are now 2 peers.
Sun Aug 11 13:55:36 2024 [info   ] ServerLobby: Message of type 1 received.
Sun Aug 11 13:55:36 2024 [verbose  ] ServerLobby: Player refused: invalid player
```

**Comment by kimden** [comment]
I made a server with 6 places on my default server machine (Ubuntu 24.04, 2 cores, 4 Gb RAM)
Tried to connect 3 local bots + 3 bots from another machine, they were fine.
Tried to connect 6 bots from another machine, they were fine.
Tried to connect 6 local bots, memory grew rapidly (+300mb per second or so), disconnecting bots didn't help indeed. Happily I didn't risk any online players and there was no one, and the other servers recovered themselves.
